### PR TITLE
dev: default to Python 3

### PR DESCRIPTION
**Pull request by joshuarli**
Supersedes https://github.com/getsentry/sentry/pull/22284, see https://github.com/getsentry/sentry/pull/22284#issuecomment-735979907. Don't merge this + friends (TODO prepare ops, getsentry PRs, and more across the board) until we have python3 via SENTRY_PYTHON3 in production.

I have a py2 `.venv` and a py3 `.venv3`. Most people will have a py2 `.venv`. We should recommend the following procedure (in an email to be drafted soon), but it's pretty much just the same easy experience as the previous email, we're just telling people to source the bootstrap script again. This is a great simplification of the procedures discussed [here](https://github.com/getsentry/sentry/pull/22284/files#r530808111). 

- upon pulling, direnv will fail
    - message updated to tell the user to bootstrap a new py3 env, see following bullet
    - **update:** warning messages added in case `.venv` is py2 and SENTRY_PYTHON2 is not set, but allow it
- create a new py3 environment - most people should only need to do this
    - `source ./scripts/bootstrap-py3-venv`: this will prompt to overwrite py2 `.venv`
- if needing py2 then create a new py2 environment (i don't feel like detecting and moving it - some venvs will break if moved etc etc)
    - `source ./scripts/bootstrap-py2-venv` just works (creates `.venv2`)
- how to switch between environments (like `export SENTRY_PYTHON2=1; direnv allow`)
- people that were previously using py3 can just remove `.venv3`

#test-getsentry

### CODE REVIEW

**Review by armenzg — APPROVED** [review]

**Comment by armenzg** [comment]
The developer that has a Python 2 venv cannot continue using their set up unless they set `SENTRY_PYTHON2` which they're not told about.

```
# Starting from a Python 2 venv
python2 -m virtualenv .venv
source ./scripts/bootstrap-py3-venv
❯ source ./scripts/bootstrap-py3-venv
You seem to have a virtualenv already, but it's Python 2. Remove it and replace with Python 3 (y/N)?
n
Aborted!
```

```
# From a Python 2 venv + env variable
sentry on  dev/py3-default [$?] via ⬢ v12.19.0 via 🐍 v2.7.16 (.venv) took 8s
❯ export SENTRY_PYTHON2=1

sentry on  dev/py3-default [$?] via ⬢ v12.19.0 via 🐍 v2.7.16 (.venv)
❯ deactivate

sentry on  dev/py3-default [$?] via ⬢ v12.19.0 via 🐍 v3.6.10
❯ direnv allow
direnv: loading ~/code/sentry/.envrc
direnv: Configuring git...
direnv: Running in DEPRECATED Python 2 mode...
direnv: Activating virtualenv...
direnv: Checking pre-commit...
direnv: Checking node...
direnv: SUCCESS!

direnv tooling is in an ALPHA state!
If you're having trouble, or have questions, please ask in #discuss-dev-tooling
and/or reach out to @josh.

direnv: export +NODE_OPTIONS +PIP_DISABLE_PIP_VERSION_CHECK +PYTHONDONTWRITEBYTECODE +SENTRY_UI_HOT_RELOAD +VIRTUAL_ENV ~PATH ~venv_name
```

**Comment by joshuarli** [comment]
Gonna leave it to @armenzg here to add warning messages to envrc and ensure-venv if `.venv` is py2 and `SENTRY_PYTHON2` is not set, but allow it.

**Comment by armenzg** [comment]
Steps to try from an existing Python 2 .env (in both cases the user needs to acknowledge removing the .venv):

```
❯ python2 -m virtualenv .venv
❯ direnv allow
# The user will see:
# To set up Python 3, run the following: source ./scripts/bootstrap-py3-venv
# To keep using Python 2, run: source ./scripts/bootstrap-py2-venv
❯ source ./scripts/bootstrap-py2-venv
You seem to have a virtualenv already. This script is meant to start one from scratch. Continue (y/N)?
```

If the user decides to go straight to Python 3 then the bootstrap script requests confirmation to remove the .venv:
```
❯ source ./scripts/bootstrap-py3-venv
You seem to have a virtualenv already. This script is meant to start one from scratch. Continue (y/N)?
y
Okay, let's do this.
```

If the user has a .venv with Python 3 version greater than 3.6 it will also work.

File: .envrc
Comment (joshuarli) [inline .envrc:158]:
```suggestion
info "Ensuring proper virtualenv..."
```

File: .envrc
Comment (joshuarli) [inline .envrc:159]: Oh, this is good deduplication.

File: scripts/ensure-venv.sh
Comment (joshuarli) [inline scripts/ensure-venv.sh:25]: Could we please keep this fast? Since it's called by direnv now. We're invoking 2 python interpreters and executing python code each time.

Just parse `python -V`. Like this:

    python -V | sed 's/\./ /g' | read _ python_major python_minor python_patch

File: scripts/bootstrap-py3-venv
Comment (EvanPurkhiser) [inline scripts/bootstrap-py3-venv:32]: Can we just change it to

```
You have an existing virtualenv. This script will OVERWRITE this. Continue (y/N)?
```

File: setup.py
Comment (joshuarli) [inline setup.py:18]: Rather than pull in logging, we could just print to sys.stderr.

File: scripts/bootstrap-py3-venv
Comment (EvanPurkhiser) [inline scripts/bootstrap-py3-venv:37]:
```
You have an existing Python 2 virtualenv. Remove it and replace with a Python 3 virtualenv? (y/N)
```

File: scripts/bootstrap-py3-venv
Comment (joshuarli) [reply scripts/bootstrap-py3-venv:32]: +1.

Comment (joshuarli) [reply scripts/bootstrap-py3-venv:32]: Oh, `OVERWRITE it` would be better.

File: scripts/bootstrap-py3-venv
Comment (joshuarli) [reply scripts/bootstrap-py3-venv:37]: Recommended to use the suggested changes feature, it also pretty prints a diff.

File: setup.py
Comment (joshuarli) [inline setup.py:14]: Let's avoid doing things like `version[0] == 3` since this wouldn't be good if Python 4 came out.

You can perform arithmetic comparisons on the entire sys.version_info tuple.

File: scripts/ensure-venv.sh
Comment (joshuarli) [inline scripts/ensure-venv.sh:28]: TIL this works, but let's just be consistent - use `"$major"`. Applies to some other places as well.

File: scripts/ensure-venv.sh
Comment (EvanPurkhiser) [inline scripts/ensure-venv.sh:30]: I thought we were going to keep the environment working? So should we not continue and just update the `python_version`?

File: setup.py
Comment (EvanPurkhiser) [reply setup.py:18]: I would be OK if we pulled in the logger just for this bit

File: setup.py
Comment (EvanPurkhiser) [inline setup.py:18]:
```suggestion
    if version[1] >= 6:
    	import logging

		logger = logging.getLogger()
        logger.warning("A Python version different than 3.6 is being used ({})".format(version[:2]))
```

File: .envrc
Comment (armenzg) [reply .envrc:159]: Thank you Josh!

File: setup.py
Comment (armenzg) [reply setup.py:21]: IMHO, I believe it is more appropriate to use logging's warning than stderr since it is a warning rather than an error.

Comment (armenzg) [reply setup.py:21]: I'm curious, what's the adversity of importing a standard library module?

File: setup.py
Comment (armenzg) [reply setup.py:11]: Thanks for catching this!

File: setup.py
Comment (armenzg) [reply setup.py:18]: FYI I put the import within the condition as Evan suggested.

File: scripts/ensure-venv.sh
Comment (armenzg) [reply scripts/ensure-venv.sh:25]: Surprisingly, the variables do not become available for the rest of the script.
I have reversed it for now. If you want to investigate it please go ahead and open a follow up PR.

File: scripts/ensure-venv.sh
Comment (armenzg) [reply scripts/ensure-venv.sh:30]: I have addressed all comments on the PR.

What about this?
- To use your existing .venv with Python simply set SENTRY_PYTHON2 and re-run
- To set up Python 3, run the following: source ./scripts/bootstrap-py3-venv
- To keep using Python 2, run: source ./scripts/bootstrap-py2-venv

I think I need to make the first option evident in the choices.

@evanh add the first option explicitly and call it done?

File: scripts/ensure-venv.sh
Comment (joshuarli) [reply scripts/ensure-venv.sh:25]: Yeah, I got this. Sorry, the reason is pipes involve subshells, so those variables won't make it to the parent shell.

The only way I can think of right now is to sed into a file to persist that data. Then `read` it in the parent shell. But there's surely a better way?

File: setup.py
Comment (joshuarli) [reply setup.py:18]: > I'm curious, what's the adversity of importing a standard library module?

Nothing much, I just tend to err on the side of being pretty conservative with software by default.

File: scripts/ensure-venv.sh
Comment (joshuarli) [reply scripts/ensure-venv.sh:25]: Gonna resolve this and follow up in the future.

File: scripts/ensure-venv.sh
Comment (joshuarli) [reply scripts/ensure-venv.sh:30]: No, if .venv is py2 then it should just continue to work, but with warnings.

Comment (EvanPurkhiser) [reply scripts/ensure-venv.sh:30]: Yeah +1 for having everything just continue to work once we merge this

**Comment by armenzg** [comment]
@EvanPurkhiser @joshuarli This is ready to go. The scripts won't stop because they're using their original py27/.venv directory.

**Comment by joshuarli** [comment]
I've just moved the colorized warning to ensure-venv only so it's not printed twice in direnv.

image

File: .envrc
Comment (joshuarli) [inline .envrc:128]: This doesn't make sense. We need to check if `.venv` contains Python 2. It hasn't been activated at this point. Just remove this check. This is also why I moved this to ensure-venv.sh, aside from colorization. What you had before was fine:

```
if [ "$SENTRY_PYTHON2" = "1" ]; then
    venv_name=".venv2"
    python_version="2.7"
fi
```

File: scripts/ensure-venv.sh
Comment (joshuarli) [inline scripts/ensure-venv.sh:41]: You don't have to fix this since this is all just temporary transitional code, but I want to point out here that this is the Python 4 problem again.

File: scripts/ensure-venv.sh
Comment (joshuarli) [inline scripts/ensure-venv.sh:51]: How did you arrive at this workaround? Under what circumstances was the script completing without `die` and exiting nonzero? I have a feeling it was some other factor, because this shouldn't be necessary.

Comment (armenzg) [reply scripts/ensure-venv.sh:51]: You will need to look at this output carefully (full output at the end):

First execution of `direnv allow` after commenting out the `exit 0` line.
```
direnv: Ensuring proper virtualenv...
direnv wasn't able to complete execution.
```

The second execution of `direnv allow` after adding back `exit 0` line.
```
direnv: Ensuring proper virtualenv...
direnv: Your virtualenv is activated, but sentry doesn't seem to be installed.
direnv: To install sentry, please type: make install-py-dev
```

```
❯ direnv allow
direnv: loading ~/code/sentry/.envrc
direnv: Configuring git...
direnv: Activating virtualenv...
direnv: You don't seem to have a virtualenv.
direnv: To create a virtualenv, please type: python3.6 -m virtualenv .venv
direnv wasn't able to complete execution.
You may have been given some recommendations in the error message.
Follow them, and then you'll need to redo direnv by running "direnv allow".

direnv tooling is in an ALPHA state!
If you're having trouble, or have questions, please ask in #discuss-dev-tooling
and/or reach out to @josh.

sentry on  dev/py3-default [$!] via ⬢ v12.19.0 via 🐍 v3.6.10
❯ python3.6 -m virtualenv .venv
created virtual environment CPython3.6.10.final.0-64 in 240ms
  creator CPython3Posix(dest=/Users/armenzg/code/sentry/.venv, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/Users/armenzg/Library/Application Support/virtualenv)
    added seed packages: pip==20.2.4, setuptools==50.3.2, wheel==0.35.1
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator

sentry on  dev/py3-default [$!] via ⬢ v12.19.0 via 🐍 v3.6.10
❯ direnv allow
direnv: loading ~/code/sentry/.envrc
direnv: Configuring git...
direnv: Activating virtualenv...
direnv: Ensuring proper virtualenv...
direnv wasn't able to complete execution.
You may have been given some recommendations in the error message.
Follow them, and then you'll need to redo direnv by running "direnv allow".

direnv tooling is in an ALPHA state!
If you're having trouble, or have questions, please ask in #discuss-dev-tooling
and/or reach out to @josh.

sentry on  dev/py3-default [$!] via ⬢ v12.19.0 via 🐍 v3.6.10
❯ direnv allow
direnv: loading ~/code/sentry/.envrc
direnv: Configuring git...
direnv: Activating virtualenv...
direnv: Ensuring proper virtualenv...
direnv: Your virtualenv is activated, but sentry doesn't seem to be installed.
direnv: To install sentry, please type: make install-py-dev
direnv wasn't able to complete execution.
You may have been given some recommendations in the error message.
Follow them, and then you'll need to redo direnv by running "direnv allow".
```

File: .envrc
Comment (armenzg) [reply .envrc:128]: @joshuarli It is needed to support the usage of `.venv` w/o requiring the user to manually have to set `SENTRY_PYTHON2`.

At this point, I would like to stop making more tweaks since this code can so easily be broken.

Comment (joshuarli) [reply .envrc:128]: The replacement I suggested _is needed_. I'm pointing out that the python version check you have is wrong. And this can break if python resolves to python3, so it's a bug that needs to be fixed, not a tweak.

Comment (armenzg) [reply .envrc:128]: I've tested it for all case scenarios and it does not fail.

Please let me know what your test case or how you get to that conclusion. I am happy to learn.

Comment (joshuarli) [reply .envrc:128]: Just because it works for all your test cases doesn't mean it's not a bug. I've already written down everything. Here it is again: the python check you have there is _before_ the venv is activated. So if it resolves to like, Python 3, the entire thing will be skipped and SENTRY_PYTHON2 doesn't do anything.

We want SENTRY_PYTHON2 to work in all cases. 

image

File: scripts/ensure-venv.sh
Comment (joshuarli) [reply scripts/ensure-venv.sh:51]: In the interest of time I'll look at this later, this shouldn't block.

File: .envrc
Comment (armenzg) [reply .envrc]: Thank you Josh! I was not testing that test case.

**Comment by billyvg** [comment]
#test-getsentry

**Comment by joshuarli** [comment]
This works flawlessly.

**Review by evanpurkhiser — APPROVED** [review]
Let's do it

**Comment by billyvg** [comment]
There's a bug in the getsentry<->sentry script, but the acceptance tests seem to be passing.

**Comment by armenzg** [comment]
🎉 Woohoo!

**Comment by armenzg** [comment]
I have landed the documentation changes:
https://github.com/getsentry/develop/pull/211

**Pull request merged**
